Antrea controller segfaults when connectivity to apiserver lost temporarily

**Describe the bug**
While running some tests to troubleshoot #545, I ran into the following segfault.

```
I0401 23:07:26.947629       1 controller.go:46] Starting Antrea Controller (version v0.6.0-dev-2b49324)
I0401 23:07:26.949167       1 client.go:33] No kubeconfig file was specified. Falling back to in-cluster config
I0401 23:07:27.268143       1 serving.go:319] Generated self-signed cert in-memory
I0401 23:07:27.947834       1 networkpolicy_controller.go:869] Starting NetworkPolicy controller
I0401 23:07:27.948105       1 networkpolicy_controller.go:872] Waiting for caches to sync for NetworkPolicy controller
I0401 23:07:27.948272       1 monitor.go:87] Starting Antrea Controller Monitor
I0401 23:07:27.989986       1 monitor.go:107] Waiting for node synced for Controller Monitor
I0401 23:07:27.992434       1 secure_serving.go:116] Serving securely on [::]:443
I0401 23:07:28.048773       1 networkpolicy_controller.go:877] Caches are synced for NetworkPolicy controller
I0401 23:07:28.090426       1 monitor.go:112] Caches are synced for Controller Monitor
I0401 23:08:40.552193       1 watch.go:104] The context had been canceled, stopping process
I0401 23:08:40.552550       1 watch.go:97] The input channel had been closed, stopping process
I0401 23:08:40.552630       1 watch.go:104] The context had been canceled, stopping process
E0401 23:08:44.606827       1 webhook.go:196] Failed to make webhook authorizer request: Post https://10.96.0.1:443/apis/authorization.k8s.io/v1beta1/subjectaccessreviews: write tcp 172.17.0.3:60378->10.96.0.1:443: write: broken pipe
E0401 23:08:44.612901       1 errors.go:77] Post https://10.96.0.1:443/apis/authorization.k8s.io/v1beta1/subjectaccessreviews: write tcp 172.17.0.3:60378->10.96.0.1:443: write: broken pipe
E0401 23:09:02.284043       1 webhook.go:107] Failed to make webhook authenticator request: Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out
E0401 23:09:02.284080       1 authentication.go:65] Unable to authenticate the request due to an error: [invalid bearer token, Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out]
E0401 23:09:02.284156       1 webhook.go:107] Failed to make webhook authenticator request: Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out
E0401 23:09:02.284179       1 authentication.go:65] Unable to authenticate the request due to an error: [invalid bearer token, Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out]
E0401 23:09:02.284249       1 webhook.go:107] Failed to make webhook authenticator request: Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out
E0401 23:09:02.284261       1 authentication.go:65] Unable to authenticate the request due to an error: [invalid bearer token, Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out]
W0401 23:09:02.284311       1 reflector.go:302] pkg/mod/github.com/tnqn/client-go@v0.0.0-20200330154227-d0a165c8fbd8/tools/cache/reflector.go:98: watch of *v1.Node ended with: an error on the server ("unable to decode an event from the watch stream: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out") has prevented the request from succeeding
E0401 23:09:02.284353       1 webhook.go:107] Failed to make webhook authenticator request: Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out
E0401 23:09:02.284369       1 authentication.go:65] Unable to authenticate the request due to an error: [invalid bearer token, Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out]
E0401 23:09:02.284379       1 webhook.go:107] Failed to make webhook authenticator request: Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out
E0401 23:09:02.284393       1 authentication.go:65] Unable to authenticate the request due to an error: [invalid bearer token, Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out]
E0401 23:09:02.284476       1 webhook.go:107] Failed to make webhook authenticator request: Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out
E0401 23:09:02.284486       1 webhook.go:196] Failed to make webhook authorizer request: Post https://10.96.0.1:443/apis/authorization.k8s.io/v1beta1/subjectaccessreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out
E0401 23:09:02.284489       1 authentication.go:65] Unable to authenticate the request due to an error: [invalid bearer token, Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out]
E0401 23:09:02.284570       1 errors.go:77] Post https://10.96.0.1:443/apis/authorization.k8s.io/v1beta1/subjectaccessreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out
W0401 23:09:02.284580       1 reflector.go:302] pkg/mod/github.com/tnqn/client-go@v0.0.0-20200330154227-d0a165c8fbd8/tools/cache/reflector.go:98: watch of *v1.Pod ended with: an error on the server ("unable to decode an event from the watch stream: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out") has prevented the request from succeeding
W0401 23:09:02.284355       1 reflector.go:302] pkg/mod/github.com/tnqn/client-go@v0.0.0-20200330154227-d0a165c8fbd8/tools/cache/reflector.go:98: watch of *v1.NetworkPolicy ended with: an error on the server ("unable to decode an event from the watch stream: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out") has prevented the request from succeeding
E0401 23:09:02.284606       1 webhook.go:107] Failed to make webhook authenticator request: Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out
E0401 23:09:02.284620       1 authentication.go:65] Unable to authenticate the request due to an error: [invalid bearer token, Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out]
E0401 23:09:02.284658       1 webhook.go:107] Failed to make webhook authenticator request: Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out
E0401 23:09:02.284674       1 authentication.go:65] Unable to authenticate the request due to an error: [invalid bearer token, Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out]
E0401 23:09:02.284748       1 webhook.go:107] Failed to make webhook authenticator request: Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out
E0401 23:09:02.284760       1 authentication.go:65] Unable to authenticate the request due to an error: [invalid bearer token, Post https://10.96.0.1:443/apis/authentication.k8s.io/v1beta1/tokenreviews: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out]
W0401 23:09:02.284571       1 reflector.go:302] pkg/mod/github.com/tnqn/client-go@v0.0.0-20200330154227-d0a165c8fbd8/tools/cache/reflector.go:98: watch of *v1.Namespace ended with: an error on the server ("unable to decode an event from the watch stream: read tcp 172.17.0.3:33016->10.96.0.1:443: read: connection timed out") has prevented the request from succeeding
E0401 23:09:03.293373       1 runtime.go:73] Observed a panic: &runtime.TypeAssertionError{_interface:(*runtime._type)(0x16a2aa0), concrete:(*runtime._type)(0x17620c0), asserted:(*runtime._type)(0x18cd000), missingMethod:""} (interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *v1.Pod)
goroutine 114 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x16e9520, 0xc0009da6c0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/runtime/runtime.go:69 +0x7b
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/runtime/runtime.go:51 +0x82
panic(0x16e9520, 0xc0009da6c0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/vmware-tanzu/antrea/pkg/controller/networkpolicy.(*NetworkPolicyController).deletePod(0xc00036e6c0, 0x17620c0, 0xc0009673c0)
	/antrea/pkg/controller/networkpolicy/networkpolicy_controller.go:732 +0x385
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnDelete(...)
	/go/pkg/mod/github.com/tnqn/client-go@v0.0.0-20200330154227-d0a165c8fbd8/tools/cache/controller.go:209
k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0x0, 0x18e5247, 0x1)
	/go/pkg/mod/github.com/tnqn/client-go@v0.0.0-20200330154227-d0a165c8fbd8/tools/cache/shared_informer.go:609 +0x1d6
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0, 0xc000295dd8, 0xc000525620, 0x420bf0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:284 +0x51
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/go/pkg/mod/github.com/tnqn/client-go@v0.0.0-20200330154227-d0a165c8fbd8/tools/cache/shared_informer.go:601 +0x79
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000525740)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000295f40, 0xdf8475800, 0x0, 0x430801, 0xc000a26480)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:88
k8s.io/client-go/tools/cache.(*processorListener).run(0xc00020ac00)
	/go/pkg/mod/github.com/tnqn/client-go@v0.0.0-20200330154227-d0a165c8fbd8/tools/cache/shared_informer.go:599 +0x9b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc0002836b0, 0xc000a22530)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:71 +0x59
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:69 +0x62
panic: interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *v1.Pod [recovered]
	panic: interface conversion: interface {} is cache.DeletedFinalStateUnknown, not *v1.Pod

goroutine 114 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/runtime/runtime.go:58 +0x105
panic(0x16e9520, 0xc0009da6c0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/vmware-tanzu/antrea/pkg/controller/networkpolicy.(*NetworkPolicyController).deletePod(0xc00036e6c0, 0x17620c0, 0xc0009673c0)
	/antrea/pkg/controller/networkpolicy/networkpolicy_controller.go:732 +0x385
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnDelete(...)
	/go/pkg/mod/github.com/tnqn/client-go@v0.0.0-20200330154227-d0a165c8fbd8/tools/cache/controller.go:209
k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0x0, 0x18e5247, 0x1)
	/go/pkg/mod/github.com/tnqn/client-go@v0.0.0-20200330154227-d0a165c8fbd8/tools/cache/shared_informer.go:609 +0x1d6
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0, 0xc000c5bdd8, 0xc000525620, 0x420bf0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:284 +0x51
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/go/pkg/mod/github.com/tnqn/client-go@v0.0.0-20200330154227-d0a165c8fbd8/tools/cache/shared_informer.go:601 +0x79
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000525740)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000c5bf40, 0xdf8475800, 0x0, 0x430801, 0xc000a26480)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:88
k8s.io/client-go/tools/cache.(*processorListener).run(0xc00020ac00)
	/go/pkg/mod/github.com/tnqn/client-go@v0.0.0-20200330154227-d0a165c8fbd8/tools/cache/shared_informer.go:599 +0x9b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc0002836b0, 0xc000a22530)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:71 +0x59
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190612205821-1799e75a0719/pkg/util/wait/wait.go:69 +0x62
```

**To Reproduce**
Run the e2e test `TestPodConnectivityAfterAntreaRestart` on a Kind cluster in a loop with `go test -v -run=TestPodConnectivityAfterAntreaRestart github.com/vmware-tanzu/antrea/test/e2e -provider=kind`.

**Expected**
`TestPodConnectivityAfterAntreaRestart` breaks the dataplane, but the Antrea Controller should not crash.

**Versions:**
Antrea
```
===> Version information <===
VERSION: v0.6.0-dev
GIT_SHA: 2b49324
GIT_TREE_STATE: clean
RELEASE_STATUS: unreleased
DOCKER_IMG_VERSION: v0.6.0-dev-2b49324
```

--- a/pkg/controller/networkpolicy/networkpolicy_controller.go
+++ b/pkg/controller/networkpolicy/networkpolicy_controller.go
@@ -637,8 +637,21 @@ func (n *NetworkPolicyController) updateNetworkPolicy(old, cur interface{}) {
 // deleteNetworkPolicy receives NetworkPolicy DELETED events and deletes resources
 // which can be consumed by agents to delete corresponding rules on the Nodes.
 func (n *NetworkPolicyController) deleteNetworkPolicy(old interface{}) {
+	np, ok := old.(*networkingv1.NetworkPolicy)
+	if !ok {
+		tombstone, ok := old.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Error decoding object when deleting NetworkPolicy, invalid type: %v", old)
+			return
+		}
+		np, ok = tombstone.Obj.(*networkingv1.NetworkPolicy)
+		if !ok {
+			klog.Errorf("Error decoding object tombstone when deleting NetworkPolicy, invalid type: %v", tombstone.Obj)
+			return
+		}
+	}
 	defer n.heartbeat("deleteNetworkPolicy")
-	np := old.(*networkingv1.NetworkPolicy)
+
 	klog.V(2).Infof("Processing NetworkPolicy %s/%s DELETE event", np.Namespace, np.Name)
 	key, _ := keyFunc(np)
 	oldInternalNPObj, _, _ := n.internalNetworkPolicyStore.Get(key)
@@ -728,8 +741,21 @@ func (n *NetworkPolicyController) updatePod(oldObj, curObj interface{}) {
 // deletePod retrieves all AddressGroups and AppliedToGroups which match the Pod's
 // labels and enqueues the groups key for further processing.
 func (n *NetworkPolicyController) deletePod(old interface{}) {
+	pod, ok := old.(*v1.Pod)
+	if !ok {
+		tombstone, ok := old.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Error decoding object when deleting Pod, invalid type: %v", old)
+			return
+		}
+		pod, ok = tombstone.Obj.(*v1.Pod)
+		if !ok {
+			klog.Errorf("Error decoding object tombstone when deleting Pod, invalid type: %v", tombstone.Obj)
+			return
+		}
+	}
 	defer n.heartbeat("deletePod")
-	pod := old.(*v1.Pod)
+
 	klog.V(2).Infof("Processing Pod %s/%s DELETE event, labels: %v", pod.Namespace, pod.Name, pod.Labels)
 	// Find all AppliedToGroup keys which match the Pod's labels.
 	appliedToGroupKeys := n.filterAppliedToGroupsForPod(pod)
@@ -785,8 +811,21 @@ func (n *NetworkPolicyController) updateNamespace(oldObj, curObj interface{}) {
 // deleteNamespace retrieves all AddressGroups which match the Namespace's
 // labels and enqueues the group keys for further processing.
 func (n *NetworkPolicyController) deleteNamespace(old interface{}) {
+	namespace, ok := old.(*v1.Namespace)
+	if !ok {
+		tombstone, ok := old.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Error decoding object when deleting Namespace, invalid type: %v", old)
+			return
+		}
+		namespace, ok = tombstone.Obj.(*v1.Namespace)
+		if !ok {
+			klog.Errorf("Error decoding object tombstone when deleting Namespace, invalid type: %v", tombstone.Obj)
+			return
+		}
+	}
 	defer n.heartbeat("deleteNamespace")
-	namespace := old.(*v1.Namespace)
+
 	klog.V(2).Infof("Processing Namespace %s DELETE event, labels: %v", namespace.Name, namespace.Labels)
 	// Find groups matching deleted Namespace's labels and enqueue them
 	// for further processing.

--- a/pkg/controller/networkpolicy/networkpolicy_controller_test.go
+++ b/pkg/controller/networkpolicy/networkpolicy_controller_test.go
@@ -2193,6 +2193,61 @@ func TestIPStrToIPAddress(t *testing.T) {
 	}
 }
 
+func TestDeleteFinalStateUnknownPod(t *testing.T) {
+	_, c := newController()
+	c.heartbeatCh = make(chan heartbeat, 2)
+	ns := metav1.NamespaceDefault
+	pod := getPod("p1", ns, "", "1.1.1.1", false)
+	c.addPod(pod)
+	key, _ := cache.MetaNamespaceKeyFunc(pod)
+	c.deletePod(cache.DeletedFinalStateUnknown{Key: key, Obj: pod})
+	close(c.heartbeatCh)
+	var ok bool
+	_, ok = <-c.heartbeatCh
+	assert.True(t, ok, "Missing event on channel")
+	_, ok = <-c.heartbeatCh
+	assert.True(t, ok, "Missing event on channel")
+}
+
+func TestDeleteFinalStateUnknownNamespace(t *testing.T) {
+	_, c := newController()
+	c.heartbeatCh = make(chan heartbeat, 2)
+	ns := &v1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "nsA",
+		},
+	}
+	c.addNamespace(ns)
+	c.deleteNamespace(cache.DeletedFinalStateUnknown{Key: "nsA", Obj: ns})
+	close(c.heartbeatCh)
+	var ok bool
+	_, ok = <-c.heartbeatCh
+	assert.True(t, ok, "Missing event on channel")
+	_, ok = <-c.heartbeatCh
+	assert.True(t, ok, "Missing event on channel")
+}
+
+func TestDeleteFinalStateUnknownNetworkPolicy(t *testing.T) {
+	_, c := newController()
+	c.heartbeatCh = make(chan heartbeat, 2)
+	np := &networkingv1.NetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{Namespace: "nsA", Name: "npA", UID: "uidA"},
+		Spec: networkingv1.NetworkPolicySpec{
+			PodSelector: metav1.LabelSelector{},
+			PolicyTypes: []networkingv1.PolicyType{networkingv1.PolicyTypeIngress},
+		},
+	}
+	c.addNetworkPolicy(np)
+	key, _ := cache.MetaNamespaceKeyFunc(np)
+	c.deleteNetworkPolicy(cache.DeletedFinalStateUnknown{Key: key, Obj: np})
+	close(c.heartbeatCh)
+	var ok bool
+	_, ok = <-c.heartbeatCh
+	assert.True(t, ok, "Missing event on channel")
+	_, ok = <-c.heartbeatCh
+	assert.True(t, ok, "Missing event on channel")
+}
+
 // util functions for testing.
 func getK8sNetworkPolicyPorts(proto v1.Protocol) []networkingv1.NetworkPolicyPort {
 	portNum := intstr.FromInt(80)

